File: Copy List with Random Pointer - Leetcode 138/Copy List with Random Pointer - Leetcode 138.ts
class _Node {
  val: number;
  next: _Node | null;
  random: _Node | null;

  constructor(val?: number, next?: _Node | null, random?: _Node) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
    this.random = random === undefined ? null : random;
  }
}
// Solution 1: T - O(n), S - O(n)
const copyRandomList = (head: _Node | null): _Node | null => {
  if (!head) return null;

  const oldToNew = new Map();
  let curr: _Node | null = head;

  while (curr) {
    oldToNew.set(curr, new _Node(curr.val));
    curr = curr.next;
  }

  curr = head;
  while (curr) {
    oldToNew.get(curr).next = oldToNew.get(curr.next) || null;
    oldToNew.get(curr).random = oldToNew.get(curr.random) || null;
    curr = curr.next;
  }

  return oldToNew.get(head);
};

// Solution 2: T - O(n), S - O(1)

const copyRandomListOptimize = (head: _Node | null): _Node | null => {
  if (!head) return null;

  let curr: _Node | null = head;
  while (curr) {
    const new_node = new _Node(curr.val, curr.next);
    curr.next = new_node;
    curr = new_node.next;
  }

  curr = head;
  while (curr) {
    if (curr.next) {
      curr.next.random = curr.random ? curr.random.next : null;
    }
    curr = curr.next ? curr.next.next : null;
  }

  const old_head: _Node | null = head;
  const new_head: _Node | null = head.next;
  let curr_old: _Node | null = old_head;
  let curr_new: _Node | null = new_head;

  while (curr_old && curr_new) {
    curr_old.next = curr_old.next ? curr_old.next.next : null;
    curr_new.next = curr_new.next ? curr_new.next.next : null;
    curr_old = curr_old.next;
    curr_new = curr_new.next;
  }

  return new_head;
};

File: Copy List with Random Pointer - Leetcode 138/Copy List with Random Pointer - Leetcode 138.py
class Node:
    def __init__(self,val: int, next: 'None' = None, random: 'None' = None):
        self.val = int(val)
        self.next = next
        self.random = random

def copyRandomList(head: 'Optional[Node]')->'Optional[Node]':
    curr = head
    oldToNew = {}
    while curr:
        oldToNew[curr] = Node(curr.val)
        curr = curr.next

    curr = head
    while curr:
        oldToNew[curr].next = oldToNew[curr.next]
        oldToNew[curr].random = oldToNew[curr.random]

        curr = curr.next

    return oldToNew[head]

def copyRandomListOptimize(head:Node)->Node:
    curr = head
    while curr:
       new_node = Node(curr.val, curr.next) 
       curr.next = new_node
       curr = new_node.next

    curr = head
    while curr:
        if curr.next:
            curr.random.next = curr.next.random if curr.next else None
        curr = curr.next.next

    old_head = head
    new_head = head.next
    curr_old = old_head
    curr_new = new_head

    while curr_old:
        old_head.next = old_head.next.next if old_head.next else None
        new_head.next = new_head.next.next if new_head.next else None
        curr_old = curr_old.next
        curr_new = curr_new.next
    
    return new_head
    